### PR TITLE
Etags and patching

### PR DESCRIPTION
**Pull request by nickbdyer**

### CODE REVIEW

File: src/main/java/uk/nickbdyer/httpserver/controllers/FileController.java [hunk 1]
```diff
@@ -19,4 +24,40 @@ public Response get(Request request) {
         return new Response(200, new HashMap<>(), file);
     }
 
+    @Override
+    public Response patch(Request request) {
+        if (etagsMatch(request)) {
```
Comment (felipesere) [inline src/main/java/uk/nickbdyer/httpserver/controllers/FileController.java]: I would extract the necessary data out of the request at this level and let the rest of the app not worry about "Request"

File: src/test/java/uk/nickbdyer/httpserver/controllers/ControllerTest.java [hunk 2]
```diff
@@ -98,6 +98,16 @@ public void aControllerWillRespondToAConnectRequest() {
     }
 
     @Test
+    public void aControllerWillRespondToAPatchRequest() {
+        Request getRequest = new Request(PATCH, "/");
```
Comment (felipesere) [inline src/test/java/uk/nickbdyer/httpserver/controllers/ControllerTest.java]: Copy paste error. Should be patchRequest.

File: src/test/java/uk/nickbdyer/httpserver/controllers/ControllerTest.java [hunk 3]
```diff
@@ -98,6 +98,16 @@ public void aControllerWillRespondToAConnectRequest() {
...
+        ControllerSpy controller = new ControllerSpy();
+
+        controller.execute(getRequest);
+
+        assertEquals("patch", controller.methodTriggered);
```
Comment (felipesere) [inline src/test/java/uk/nickbdyer/httpserver/controllers/ControllerTest.java:115]: Shouldn't this also be testing the 405 method not allowed?

File: src/test/java/uk/nickbdyer/httpserver/controllers/FileControllerTest.java [hunk 4]
```diff
@@ -50,4 +55,32 @@ public void willReturnFileContentsInBody() throws IOException {
         assertEquals("hello", new String(response.getBody()));
     }
 
+    @Test
+    public void willReturn204ForValidPatchEtag() throws IOException {
```
Comment (felipesere) [inline src/test/java/uk/nickbdyer/httpserver/controllers/FileControllerTest.java]: This test reinforces my argument from above re. "Too much low level in controller". 

To make sure that the sha1 matching works, I should not need requests, responses etc. this feels complicated.

Comment (nickbdyer) [reply src/test/java/uk/nickbdyer/httpserver/controllers/FileControllerTest.java]: I think that in this case the test name is wrong. I believe I should have a test for a 204 response in this controller, but perhaps I should return 204 regardless of the presence of a "If-Match" header. The test for sha1 matching I expect will be in my "FileManipulation" class, or whatever it gets called when I extract it out.

**Pull request merged**